### PR TITLE
[perf] Add unmanaged constraint to some APIs to allow pinning, and not manual GCHandle code

### PR DESCRIPTION
**Pull request by Therzok**

### CODE REVIEW

**Comment by Therzok** [comment]
Hmmm, what would be the problem with T here? This isn't public API

File: src/AudioToolbox/AudioFormatAvailability.cs [hunk 1]
```diff
@@ -57,25 +57,21 @@ public static AudioClassDescription[] GetEncoders (AudioFormatType format)
...
 
-			var data = new T[size / Marshal.SizeOf (typeof (T))];
-			var array_handle = GCHandle.Alloc (data, GCHandleType.Pinned); // This requires a pinned GCHandle, since it's not possible to use unsafe code to get the address of a generic object.
+			var data = new T[size / sizeof (T)];
 
-			try {
-				var ptr = array_handle.AddrOfPinnedObject ();
-				var res = AudioFormatPropertyNative.AudioFormatGetProperty (prop, sizeof (AudioFormatType), ref format, ref size, ptr);
+			fixed (T* ptr = data) {
+				var res = AudioFormatPropertyNative.AudioFormatGetProperty (prop, sizeof (AudioFormatType), ref format, ref size, (IntPtr)ptr);
```
Comment (rolfbjarne) [inline src/AudioToolbox/AudioFormatAvailability.cs]:
```suggestion
				var res = AudioFormatPropertyNative.AudioFormatGetProperty (prop, sizeof (AudioFormatType), ref format, ref size, (IntPtr) ptr);
```

File: src/AudioToolbox/AudioConverter.cs [hunk 2]
```diff
@@ -650,26 +650,22 @@ unsafe AudioValueRange[] GetAudioValueRange (AudioConverterPropertyID prop)
...
 				return null;
 
 			var data = new T [size / elementSize];
-			var array_handle = GCHandle.Alloc (data, GCHandleType.Pinned); // This requires a pinned GCHandle, since it's not possible to use unsafe code to get the address of a generic object.
 
-			try {
-				var ptr = array_handle.AddrOfPinnedObject ();
-				var res = AudioConverterGetProperty (handle, prop, ref size, ptr);
+			fixed (T* ptr = data) {
+				var res = AudioConverterGetProperty (handle, prop, ref size, (IntPtr)ptr);
```
Comment (rolfbjarne) [inline src/AudioToolbox/AudioConverter.cs]:
```suggestion
				var res = AudioConverterGetProperty (handle, prop, ref size, (IntPtr) ptr);
```

**Comment by spouliot** [comment]
@Therzok being `private` makes it easier, not complete. 
Did you review all usage to determine which types are used for `T` ? 
and checked they are all behaved ing identically for [un]managed size ?
If so please add the list to your commit message, thanks!

**Comment by Therzok** [comment]
Any suggestion of where to put the tests?

**Comment by spouliot** [comment]
@Therzok `monotouch-test`

**Comment by Therzok** [comment]
Updated

**Comment by Therzok** [comment]
Build didn't kick in?

**Comment by rolfbjarne** [comment]
Something funny's going on.
> Started 1 day 3 hr ago

**Comment by rolfbjarne** [comment]
build

**Comment by rolfbjarne** [comment]
build

File: src/AudioToolbox/AudioFormatAvailability.cs [hunk 3]
```diff
@@ -57,26 +57,23 @@ public static AudioClassDescription[] GetEncoders (AudioFormatType format)
...
 
-			var data = new T[size / Marshal.SizeOf (typeof (T))];
-			var array_handle = GCHandle.Alloc (data, GCHandleType.Pinned); // This requires a pinned GCHandle, since it's not possible to use unsafe code to get the address of a generic object.
+			var data = new T[size / sizeof (T)];
 
-			try {
-				var ptr = array_handle.AddrOfPinnedObject ();
-				var res = AudioFormatPropertyNative.AudioFormatGetProperty (prop, sizeof (AudioFormatType), ref format, ref size, ptr);
+			fixed (T* ptr = data) {
+				var res = AudioFormatPropertyNative.AudioFormatGetProperty (prop, sizeof (AudioFormatType), ref format, ref size, (IntPtr) ptr);
```
Comment (Therzok) [inline src/AudioToolbox/AudioFormatAvailability.cs]: Nothing should be different for this call, except using fixed vs manual pinning, this is confusing.

File: src/AudioToolbox/AudioConverter.cs [hunk 4]
```diff
@@ -650,26 +650,22 @@ unsafe AudioValueRange[] GetAudioValueRange (AudioConverterPropertyID prop)
...
 
-			try {
-				var ptr = array_handle.AddrOfPinnedObject ();
-				var res = AudioConverterGetProperty (handle, prop, ref size, ptr);
+			fixed (T* ptr = data) {
+				var res = AudioConverterGetProperty (handle, prop, ref size, (IntPtr) ptr);
 				if (res != 0)
 					return null;
 
 				Array.Resize (ref data, size / elementSize);
```
Comment (Therzok) [inline src/AudioToolbox/AudioConverter.cs]: This should probably be done outside the fixed statement.

File: src/AudioToolbox/AudioFormatAvailability.cs [hunk 5]
```diff
@@ -71,6 +71,9 @@ public static class AudioFormatAvailability {
 			if (AudioFormatPropertyNative.AudioFormatGetPropertyInfo (prop, sizeof (AudioFormatType), ref format, out size) != 0)
 				return null;
 
+			if (size == 0)
+				return Array.Empty<T> ();
```
Comment (Therzok) [inline src/AudioToolbox/AudioFormatAvailability.cs:75]: 😧

**Review by dalexsoto — APPROVED** [review]

**Review by mandel-macaque — APPROVED** [review]

**Review by tj-devel709 — APPROVED** [review]

**Pull request merged**